router/pkg/pubsub/nats/engine_datasource.go

```
@@ -9,38 +9,90 @@ import (
+}
+
+// ProviderID returns the provider ID
+func (s *SubscriptionEventConfiguration) ProviderID() string {
+	return s.Provider
+}
+
+// ProviderType returns the provider type
+func (s *SubscriptionEventConfiguration) ProviderType() datasource.ProviderType {
+	return datasource.ProviderTypeNats
+}
+
+// RootFieldName returns the root field name
+func (s *SubscriptionEventConfiguration) RootFieldName() string {
+	return s.FieldName
 }
 
 type PublishAndRequestEventConfiguration struct {
-	ProviderID string          `json:"providerId"`
-	Subject    string          `json:"subject"`
-	Data       json.RawMessage `json:"data"`
+	Provider  string `json:"providerId"`
+	Subject   string `json:"subject"`
+	Event     Event  `json:"event"`
+	FieldName string `json:"rootFieldName"`
+}
+
+// ProviderID returns the provider ID
+func (p *PublishAndRequestEventConfiguration) ProviderID() string {
+	return p.Provider
 }
 
-func (s *PublishAndRequestEventConfiguration) MarshalJSONTemplate() string {
+// ProviderType returns the provider type
+func (p *PublishAndRequestEventConfiguration) ProviderType() datasource.ProviderType {
+	return datasource.ProviderTypeNats
+}
+
+// RootFieldName returns the root field name
+func (p *PublishAndRequestEventConfiguration) RootFieldName() string {
+	return p.FieldName
+}
+
+func (s *PublishAndRequestEventConfiguration) MarshalJSONTemplate() (string, error) {
 	// The content of the data field could be not valid JSON, so we can't use json.Marshal
 	// e.g. {"id":$$0$$,"update":$$1$$}
-	return fmt.Sprintf(`{"subject":"%s", "data": %s, "providerId":"%s"}`, s.Subject, s.Data, s.ProviderID)
+	return fmt.Sprintf(`{"subject":"%s", "event": {"data": %s}, "providerId":"%s"}`, s.Subject, s.Event.Data, s.ProviderID()), nil
 }
 
 type SubscriptionSource struct {
 	pubSub Adapter
 }
 
+func (s *SubscriptionSource) SubscriptionEventConfiguration(input []byte) datasource.SubscriptionEventConfiguration {
```

Same here, silent error.

---

fixed!